### PR TITLE
resolves #131 remove HTML tags and entites in the PDF outline

### LINKED ISSUE #131: Encoding of & in pdf bookmarks
Using alpha 4:

When using an ampersand, `&` in a section title we get the escaped value `&amp;` in the table of contents for the section, when the correct value is `&`

See screenshot of this behaviour:

image

The corresponding asciidoc might be, for example:

```adoc
[data-type="issue",data-groups="circuit,added",data-date-created="01/07/2018",data-date-modified="01/07/2018"]
=== C:015 R&I Diagrams in Standard Packages

Where R&I diagrams are provided in a standard package but do not cover every SLD orientation (e.g. top to bottom as well as bottom to top for given voltage levels in the transformer package) it is expected that the R&I diagram should be flipped in AutoCAD.
```

### PR DESCRIPTION
**Pull request by Mogztter**
resolves #131

### CODE REVIEW

File: lib/outline.js [hunk 1]
```diff
@@ -4,14 +4,22 @@ function getOutline (node, depth) {
...
-
   return node.getSections()
     .filter(section => section.getId())
-    .map(section => ({
-      title: section.getTitle(),
-      destination: section.getId(),
-      children: getOutline(section, depth - 1)
-    }))
+    .map(section => {
+      const title = section.getTitle()
```
Comment (Mogztter) [inline lib/outline.js]: the list of replacements is not complete

Comment (danyill) [reply lib/outline.js]: Would you consider using a [package](https://www.npmjs.com/package/html-entities)?

Comment (danyill) [reply lib/outline.js]: BTW, thanks so much for looking into this.

Comment (Mogztter) [reply lib/outline.js]: I don't mind using the popular and dependency-free `html-entities` package but we will also need to encode the string to conform with the specification:

> Certain strings contain information that is intended to be human-readable, such
as text annotations, bookmark names, article names, document information, and
so forth. Such strings are referred to as text strings. Text strings are encoded in
either PDFDocEncoding or Unicode character encoding. PDFDocEncoding is a
superset of the ISO Latin 1 encoding and is documented in Appendix D. Unicode
is described in the Unicode Standard by the Unicode Consortium (see the Bibli-
ography).
For text strings encoded in Unicode, the first two bytes must be 254 followed by
255, representing the Unicode byte order marker, U+FEFF . (This sequence con-
flicts with the PDFDocEncoding character sequence thorn ydieresis , which is un-
likely to be a meaningful beginning of a word or phrase.) The remainder of the
string consists of Unicode character codes, according to the UTF-16 encoding
specified in the Unicode standard, version 2.0. Commonly used Unicode values
are represented as 2 bytes per character, with the high-order byte appearing first
in the string.

https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/pdf_reference_archives/PDFReference.pdf

Comment (Mogztter) [reply lib/outline.js]: I thought the text wasn't properly encoded but it appears that the right double quotation mark is not rendered properly (at least on my machine): https://www.fileformat.info/info/unicode/char/201d/index.htm

typo-quote-ide

I've tried to open the PDF in both [Evince](https://wiki.gnome.org/Apps/Evince) (the default Gnome PDF viewer) and Okular:


Evince | Okular
------------ | -------------
typo-quote | okular

Comment (Mogztter) [reply lib/outline.js]: It looks fine when using Asciidoctor PDF (Ruby):

asciidoctor-pdf-ruby

Comment (Mogztter) [reply lib/outline.js]: I think I got it!
I'm now using `PDFHexString` to encode the string and I'm using a function to sanitize the string similar to what Asciidoctor PDF (Ruby) is doing:


```ruby
module Asciidoctor
module PDF
module Sanitizer
  (BuiltInNamedEntities = {
    'amp' => ?&,
    'apos' => ?',
    'gt' => ?>,
    'lt' => ?<,
    'nbsp' => ' ',
    'quot' => ?",
  }).default = ??
  SanitizeXMLRx = /<[^>]+>/
  CharRefRx = /&(?:([a-z][a-z]+\d{0,2})|#(?:(\d\d\d{0,4})|x([a-f\d][a-f\d][a-f\d]{0,3})));/

  # Strip leading, trailing and repeating whitespace, remove XML tags and
  # resolve all entities in the specified string.
  def sanitize string
    string = string.gsub(SanitizeXMLRx, '') if string.include? '<'
    string = string.gsub(CharRefRx) { $1 ? BuiltInNamedEntities[$1] : ([$2 ? $2.to_i : ($3.to_i 16)].pack 'U1') } if string.include? '&'
    string.strip.tr_s ' ', ' '
  end
end
```

File: lib/outline.js [hunk 2]
```diff
@@ -1,17 +1,50 @@
...
+  'lt': '<',
+  'nbsp': ' ',
+  'quot': '"',
+}
+const SanitizeXMLRx = /<[^>]+>/g
+const CharRefNamedRx = /&([a-z][a-z]+\d{0,2});/g
+const CharRefDecimalRx = /&#(\d\d\d{0,4});/g
+const CharRefHexRx = /&#x([a-z\d][a-z\d][a-z\d]{0,3});/g
+
+function sanitize (string) {
```
Comment (Mogztter) [inline lib/outline.js]: I think we can replace this whole function with:

```js
const Entities = require('html-entities').AllHtmlEntities
const entities = new Entities()

function sanitize (string) {
  entities.decode(string)
}
```

Using https://www.npmjs.com/package/html-entities as suggested by @danyill

**Pull request merged**